libs/cardano-data/src/Data/OSet/Strict.hs

```diff
@@ -0,0 +1,209 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Data.OSet.Strict (
+  OSet (Empty, (:<|:), (:|>:)),
+  null,
+  size,
+  empty,
+  singleton,
+  lookup,
+  member,
+  (!?),
+  fromStrictSeq,
+  fromStrictSeqDuplicates,
+  toStrictSeq,
+  fromSet,
+  invariantHolds,
+  invariantHolds',
+  (|>),
+  (<|),
+  (|><),
+  (><|),
+)
+where
+
+import Cardano.Ledger.Binary (DecCBOR (decCBOR), EncCBOR (encCBOR))
+import Cardano.Ledger.TreeDiff (ToExpr)
+import Control.DeepSeq (NFData)
+import Data.Foldable (foldl')
+import Data.Sequence.Strict qualified as SSeq
+import Data.Set qualified as Set
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks)
+import Prelude hiding (lookup, null, seq)
+
+-- | A general-purpose finite, ordered, set that is strict in its
+-- values.
+--
+-- The strictness is enforced by the underlying `StrictSeq` from base,
+-- and the uniqueness of the values is enforced in this module, by the
+-- constructing functions by leveraging an accompanying `Set`.
+data OSet a = OSet
+  { osSSeq :: !(SSeq.StrictSeq a)
+  , osSet :: !(Set.Set a)
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (ToExpr, NoThunks, NFData)
+
```

Comment:
Let's also add a Foldble instance. It should be very simple to do. All that is necessary for all functions in Foldable to be used with `osSSeq`, eg.
```haskell
  foldl f acc = foldl f acc . osSSeq